### ISSUE TITLE
The head was initialized incorrectly

### ISSUE THREAD

**Issue by Dasadaly** [roboflow/rf-detr#509]
### Search before asking

- [x] I have searched the RF-DETR issues and found no similar bug report.

### Bug

Fine-tuning rf-detr on 2 classes, still took its head from the standard coco config with class 80, why?
`transformer.enc_out_class_embed.0.weight torch.Size([91, 256])`
But at the same time in the arguments(config) of the model it is num_classes=2, and 2 class names class_names=[
'keyboard', 'computer mouse'].
I also shifted the classes from 1 and 2 to 0, 2 in my annotations.
the results were quite good, but when I started using the model, it did not find the objects I needed, but found completely different objects, by type of car, etc.

<img width="1800" height="1200" alt="Image" src="https://github.com/user-attachments/assets/637ef12b-84ad-4820-8dfe-8156ea1df8df" />

### Environment

- RF-DETR: 1.3.0
- OS: Ubuntu 24.04.3 LTS
- Python: 3.12.3
- Pytorch: 2.8.0
- CUDA: V12.6.85
- GPU: RTX 4060 TI (16GB)

### Minimal Reproducible Example

```
model = RFDETRMedium(num_classes=2)
model.train(
    dataset_dir="s/o/m/e/p/a/t/h",
    epochs=80,
    batch_size=1,
    grad_accum_steps=8,
    lr=1e-4,
    output_dir="result",
    resolution=2688,
    gradient_checkpointing=True
)
```

### Are you willing to submit a PR?

- [ ] Yes, I'd like to help by submitting a PR!

**Comment by devVipin01** [comment]
I also have same isssue, I train model on single class but at the infer time i am seeing COCO clases
###  Below is Training Log

Namespace(num_classes=1, grad_accum_steps=8, amp=True, lr=0.0001, lr_encoder=0.00015, batch_size=2, weight_decay=0.0001, epochs=65, lr_drop=100, clip_max_norm=0.1, lr_vit_layer_decay=0.8, lr_component_decay=0.7, do_benchmark=False, dropout=0, drop_path=0.0, drop_mode='standard', drop_schedule='constant', cutoff_epoch=0, pretrained_encoder=None, pretrain_weights='rf-detr-base.pth', pretrain_exclude_keys=None, pretrain_keys_modify_to_load=None, pretrained_distiller=None, encoder='dinov2_windowed_small', vit_encoder_num_layers=12, window_block_indexes=None, position_embedding='sine', out_feature_indexes=[2, 5, 8, 11], freeze_encoder=False, layer_norm=True, rms_norm=False, backbone_lora=False, force_no_pretrain=False, dec_layers=3, dim_feedforward=2048, hidden_dim=256, sa_nheads=8, ca_nheads=16, num_queries=300, group_detr=13, two_stage=True, projector_scale=['P4'], lite_refpoint_refine=True, num_select=300, dec_n_points=2, decoder_norm='LN', bbox_reparam=True, freeze_batch_norm=False, set_cost_class=2, set_cost_bbox=5, set_cost_giou=2, cls_loss_coef=1.0, bbox_loss_coef=5, giou_loss_coef=2, focal_alpha=0.25, aux_loss=True, sum_group_losses=False, use_varifocal_loss=False, use_position_supervised_loss=False, ia_bce_loss=True, dataset_file='roboflow', coco_path=None, dataset_dir='C:\\Users\\vipin\\Downloads\\data', square_resize_div_64=True, output_dir='C:\\Users\\vipin\\Downloads\\data\\weights', dont_save_weights=False, checkpoint_interval=10, seed=42, resume='', start_epoch=0, eval=False, use_ema=True, ema_decay=0.993, ema_tau=100, num_workers=0, device='cuda', world_size=1, dist_url='env://', sync_bn=True, fp16_eval=False, encoder_only=False, backbone_only=False, resolution=560, use_cls_token=False, multi_scale=True, expanded_scales=True, do_random_resize_via_padding=False, warmup_epochs=0.0, lr_scheduler='step', lr_min_factor=0.0, early_stopping=False, early_stopping_patience=10, early_stopping_min_delta=0.001, early_stopping_use_ema=False, gradient_checkpointing=False, patch_size=14, num_windows=4, positional_encoding_size=37, mask_downsample_ratio=4, tensorboard=True, wandb=False, project=None, run=None, class_names=['defect'], run_test=True, segmentation_head=False, distributed=False)
number of params: 31850710
[840]
loading annotations into memory...
Done (t=0.00s)
creating index...
index created!
[840]
loading annotations into memory...
Done (t=0.00s)
creating index...
index created!
[840]
loading annotations into memory...
Done (t=0.00s)
creating index...
index created!
Get benchmark
Start training
Grad accum steps:  8
Total batch size:  16
LENGTH OF DATA LOADER: 7

### Below is output model.class_names after trainng

{1: 'person', 2: 'bicycle', 3: 'car', 4: 'motorcycle', 5: 'airplane', 6: 'bus', 7: 'train', 8: 'truck', 9: 'boat', 10: 'traffic light', 11: 'fire hydrant', 13: 'stop sign', 14: 'parking meter', 15: 'bench', 16: 'bird', 17: 'cat', 18: 'dog', 19: 'horse', 20: 'sheep', 21: 'cow', 22: 'elephant', 23: 'bear', 24: 'zebra', 25: 'giraffe', 27: 'backpack', 28: 'umbrella', 31: 'handbag', 32: 'tie', 33: 'suitcase', 34: 'frisbee', 35: 'skis', 36: 'snowboard', 37: 'sports ball', 38: 'kite', 39: 'baseball bat', 40: 'baseball glove', 41: 'skateboard', 42: 'surfboard', 43: 'tennis racket', 44: 'bottle', 46: 'wine glass', 47: 'cup', 48: 'fork', 49: 'knife', 50: 'spoon', 51: 'bowl', 52: 'banana', 53: 'apple', 54: 'sandwich', 55: 'orange', 56: 'broccoli', 57: 'carrot', 58: 'hot dog', 59: 'pizza', 60: 'donut', 61: 'cake', 62: 'chair', 63: 'couch', 64: 'potted plant', 65: 'bed', 67: 'dining table', 70: 'toilet', 72: 'tv', 73: 'laptop', 74: 'mouse', 75: 'remote', 76: 'keyboard', 77: 'cell phone', 78: 'microwave', 79: 'oven', 80: 'toaster', 81: 'sink', 82: 'refrigerator', 84: 'book', 85: 'clock', 86: 'vase', 87: 'scissors', 88: 'teddy bear', 89: 'hair drier', 90: 'toothbrush'}

**Comment by Dasadaly** [comment]
> I also have same isssue, I train model on single class but at the infer time i am seeing COCO clases
> 
> ### Below is Training Log
> Namespace(num_classes=1, grad_accum_steps=8, amp=True, lr=0.0001, lr_encoder=0.00015, batch_size=2, weight_decay=0.0001, epochs=65, lr_drop=100, clip_max_norm=0.1, lr_vit_layer_decay=0.8, lr_component_decay=0.7, do_benchmark=False, dropout=0, drop_path=0.0, drop_mode='standard', drop_schedule='constant', cutoff_epoch=0, pretrained_encoder=None, pretrain_weights='rf-detr-base.pth', pretrain_exclude_keys=None, pretrain_keys_modify_to_load=None, pretrained_distiller=None, encoder='dinov2_windowed_small', vit_encoder_num_layers=12, window_block_indexes=None, position_embedding='sine', out_feature_indexes=[2, 5, 8, 11], freeze_encoder=False, layer_norm=True, rms_norm=False, backbone_lora=False, force_no_pretrain=False, dec_layers=3, dim_feedforward=2048, hidden_dim=256, sa_nheads=8, ca_nheads=16, num_queries=300, group_detr=13, two_stage=True, projector_scale=['P4'], lite_refpoint_refine=True, num_select=300, dec_n_points=2, decoder_norm='LN', bbox_reparam=True, freeze_batch_norm=False, set_cost_class=2, set_cost_bbox=5, set_cost_giou=2, cls_loss_coef=1.0, bbox_loss_coef=5, giou_loss_coef=2, focal_alpha=0.25, aux_loss=True, sum_group_losses=False, use_varifocal_loss=False, use_position_supervised_loss=False, ia_bce_loss=True, dataset_file='roboflow', coco_path=None, dataset_dir='C:\Users\vipin\Downloads\data', square_resize_div_64=True, output_dir='C:\Users\vipin\Downloads\data\weights', dont_save_weights=False, checkpoint_interval=10, seed=42, resume='', start_epoch=0, eval=False, use_ema=True, ema_decay=0.993, ema_tau=100, num_workers=0, device='cuda', world_size=1, dist_url='env://', sync_bn=True, fp16_eval=False, encoder_only=False, backbone_only=False, resolution=560, use_cls_token=False, multi_scale=True, expanded_scales=True, do_random_resize_via_padding=False, warmup_epochs=0.0, lr_scheduler='step', lr_min_factor=0.0, early_stopping=False, early_stopping_patience=10, early_stopping_min_delta=0.001, early_stopping_use_ema=False, gradient_checkpointing=False, patch_size=14, num_windows=4, positional_encoding_size=37, mask_downsample_ratio=4, tensorboard=True, wandb=False, project=None, run=None, class_names=['defect'], run_test=True, segmentation_head=False, distributed=False) number of params: 31850710 [840] loading annotations into memory... Done (t=0.00s) creating index... index created! [840] loading annotations into memory... Done (t=0.00s) creating index... index created! [840] loading annotations into memory... Done (t=0.00s) creating index... index created! Get benchmark Start training Grad accum steps: 8 Total batch size: 16 LENGTH OF DATA LOADER: 7
> 
> ### Below is output model.class_names after trainng
> {1: 'person', 2: 'bicycle', 3: 'car', 4: 'motorcycle', 5: 'airplane', 6: 'bus', 7: 'train', 8: 'truck', 9: 'boat', 10: 'traffic light', 11: 'fire hydrant', 13: 'stop sign', 14: 'parking meter', 15: 'bench', 16: 'bird', 17: 'cat', 18: 'dog', 19: 'horse', 20: 'sheep', 21: 'cow', 22: 'elephant', 23: 'bear', 24: 'zebra', 25: 'giraffe', 27: 'backpack', 28: 'umbrella', 31: 'handbag', 32: 'tie', 33: 'suitcase', 34: 'frisbee', 35: 'skis', 36: 'snowboard', 37: 'sports ball', 38: 'kite', 39: 'baseball bat', 40: 'baseball glove', 41: 'skateboard', 42: 'surfboard', 43: 'tennis racket', 44: 'bottle', 46: 'wine glass', 47: 'cup', 48: 'fork', 49: 'knife', 50: 'spoon', 51: 'bowl', 52: 'banana', 53: 'apple', 54: 'sandwich', 55: 'orange', 56: 'broccoli', 57: 'carrot', 58: 'hot dog', 59: 'pizza', 60: 'donut', 61: 'cake', 62: 'chair', 63: 'couch', 64: 'potted plant', 65: 'bed', 67: 'dining table', 70: 'toilet', 72: 'tv', 73: 'laptop', 74: 'mouse', 75: 'remote', 76: 'keyboard', 77: 'cell phone', 78: 'microwave', 79: 'oven', 80: 'toaster', 81: 'sink', 82: 'refrigerator', 84: 'book', 85: 'clock', 86: 'vase', 87: 'scissors', 88: 'teddy bear', 89: 'hair drier', 90: 'toothbrush'}

Did you use the medium model?

**Comment by TratsiakY** [comment]
Confirm! I have the same issue with at leat small and nano segmentation models

**Comment by TratsiakY** [comment]
I want to clarify. The issue is here:

In RFDETR class

```
def train_from_config(self, config: TrainConfig, **kwargs):
    ......
    elif config.dataset_file == "coco":
        class_names = COCO_CLASSES
        num_classes = 90
    else:
        raise ValueError(f"Invalid dataset file: {config.dataset_file}")
```

and later

```
@property
    def class_names(self):
        if hasattr(self.model, 'class_names') and self.model.class_names:
            return {i+1: name for i, name in enumerate(self.model.class_names)}
        return COCO_CLASSES
```

It looks like a bit strange solution )) 

And also a bit questionable moment in build_model() with string 

`load_dinov2_weights=args.pretrain_weights is None`

especially when it originates from config 

`pretrain_weights: Optional[str] = None`
